### PR TITLE
Wrappers evolution

### PR DESCRIPTION
**Pull request by gkoz**
* Add `Refcounted` wrapper.
 * Implement `Object` wrapper via `Refcounted`.
 * Change placement of the lifetime in `ToGlibPtr` for saner implementations.
 * Shift some implementations burden from generics to macros.
 * Upcasts are implicit now.

"Reformed" branches: https://github.com/gtk-rs/gdk/pull/84 and [gtk](https://github.com/gkoz/gtk-rs-gtk/tree/object_reform_gir_wrappers2)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 language: rust
 rust:
-  - beta
+  - nightly
```
Comment (GuillaumeGomez) [inline .travis.yml]: Maybe we should just add beta too once and for all. Like this, all three stages will be tested. That's not very useful but whatever...

**Comment by GuillaumeGomez** [comment]
Except for build targets, all seems ok. That's a very impressive work, thanks a lot for this !

File: .travis.yml (hunk 1 above)
Comment (GuillaumeGomez) [inline .travis.yml]: Same here.

Comment (gkoz) [inline .travis.yml]: Yeah I guess there's no harm in having beta too.

**Comment by GuillaumeGomez** [comment]
Perfect, examples will need to be updated too. Wanna do it ?

**Comment by gkoz** [comment]
I don't think examples are affected, the build breaks because the other two PRs and this one are mutually dependent.

**Comment by GuillaumeGomez** [comment]
Ok, then let's merge ! :D

**Pull request merged**